langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/ToolExecutionHelper.java

```
@@ -21,24 +20,24 @@ static String extractResult(JsonNode result) {
             JsonNode resultNode = result.get("result");
             if (resultNode.has("content")) {
                 String content = extractSuccessfulResult((ArrayNode) resultNode.get("content"));
-                boolean isError = false;
-                if (resultNode.has("isError")) {
-                    isError = resultNode.get("isError").asBoolean();
-                }
-                if (isError) {
-                    content = String.format(EXECUTION_ERROR_MESSAGE + ". The tool returned: %s", content);
+                if (isError(resultNode)) {
+                    throw new ToolExecutionException(new RuntimeException(content));
```

This double exception looks wasteful, can't we simply add one more constructor to the `ToolExecutionException` taking the `content` as argument?

---

`ToolExecutionException` will be unwrapped and error handler will see only internal `RuntimeException`. Otherwise it will be inconsistent, some exceptions will be actual exceptions, but in this case it will be a `ToolExecutionException` which is used only as an API between `ToolExecutor` and AI Service/error handler...

---

Ok, I see. At least you could still add another constructor to the `ToolExecutionException` and create the necessary `RuntimeException` only there, thus hiding this detail here and in all other places where you're doing the same. This will help also if in future we will decide that we don't need that internal `RuntimeException` anymore and we want to propagate the exception in a different way (like propagating the `ToolExecutionException` itself if it has no cause), but this is a very minor improvement, so feel free to ignore it.

---

@mariofusco good point, let me do the change